Generate CamelCase function names instead of under_score to pass golint check

Barberrrry opened the pull request:
Generate CamelCase function names instead of under_score according to default Go naming convention. 

I changed also some generated variable names and added generated comment before exported functions.

The goal of changes is to generate the code which will pass default golint/gohint check without adding easyjson-generated files to ignore.

Barberrrry commented:
Guys, any feedback on this PR?

vstarodub commented:
Hi! Sorry for the delay.

Why not just add "//golint:ignore" on the generated file?

Barberrrry commented:
I can't find such option for golint. Could you share a link, where did you see that?

vstarodub commented:
Sorry, it seems that I've found it in a commit from a rejected PR: https://github.com/golang/lint/pull/148 

It seems though that the golint developers explicitly do not want people to use it automatically on everything.

Barberrrry commented:
So, it would be great if generated code can pass golint check "from the box".

Barberrrry commented:
@vstarodub, so what's you decision about current PR?

vstarodub commented:
Sorry for such a delay, I've finally had some time to take a closer look at the PR. If you're still up for it, here are my concerns.

I think a better approach would be to always generate functions in camel case. There doesn't seem to be any need for preserving backward compatibility for internal functions. The code would be much simpler.

Also, a golint test would be nice, so that we can be sure that the final goal is acheived.

Barberrrry commented:
Thank you for your feedback. I've removed `-camel_case_functions` flag and changed default generation to use CamelCase function namer. Also, I've added `golint` check to `make test`.

vstarodub commented on easyjson/main.go:
```diff
@@ -40,13 +40,13 @@ func generate(fname string) (err error) {
 		PkgName:         p.PkgName,
 		Types:           p.StructNames,
-		SnakeCase:       *snakeCase,
+		SnakeCaseFields: *snakeCaseFields,
```
Do you mind renaming it back?

vstarodub commented on gen/generator.go:
```diff
@@ -374,3 +383,23 @@ func (SnakeCaseFieldNamer) GetJSONFieldName(t reflect.Type, f reflect.StructFiel
+// CamelCaseFunctionNamer implements FunctionNamer interface with CamelCase format
+type CamelCaseFunctionNamer struct{}
+
+func (CamelCaseFunctionNamer) GetName(keepFirst bool, parts ...string) string {
```
I'd make it a simple function (not a function on a struct) and call it directly rather than via an interface. I don't think we'll ever need any extensibility in naming auxiliary internal functions.

vstarodub commented:
Thanks for the fix!
